File: scripts/replace_urls.py
import os
import re
import glob

# Step 1: Extract all calculator slugs
slugs = set()
data_dir = r"d:\MegaCalc-Hub\src\data\calculators"
for filepath in glob.glob(os.path.join(data_dir, "*.ts")):
    with open(filepath, "r", encoding="utf-8") as f:
        content = f.read()
        # Find all slug: 'something' or slug: "something"
        matches = re.findall(r"slug\s*:\s*['\"]([^'\"]+)['\"]", content)
        for m in matches:
            slugs.add(m)

print(f"Found {len(slugs)} calculator slugs.")

# Step 2: Global replace in files
src_dir = r"d:\MegaCalc-Hub\src"
files_to_check = []
for ext in ["*.ts", "*.tsx", "*.js", "*.jsx"]:
    files_to_check.extend(glob.glob(os.path.join(src_dir, "**", ext), recursive=True))

count = 0
for filepath in files_to_check:
    try:
        with open(filepath, "r", encoding="utf-8") as f:
            content = f.read()
            original_content = content
        
        # We look for /category/<anything>/slug or /category/<anything>/<anything>/slug
        for slug in slugs:
            # Match href="/category/anything/slug" or href={`/category/anything/slug`}
            # We will just replace exactly "/category/.../slug" -> "/slug"
            # It handles both 1 or 2 level deep paths after /category/
            # E.g. /category/sports-training/bowling-average-calculator
            # E.g. /category/education/maths/fractions-calculator
            pattern = r"/category/[^/\"\'`<>]+/" + slug + r"(?=[/\"\'`<>#?])"
            content = re.sub(pattern, f"/{slug}", content)
            
            pattern2 = r"/category/[^/\"\'`<>]+/[^/\"\'`<>]+/" + slug + r"(?=[/\"\'`<>#?])"
            content = re.sub(pattern2, f"/{slug}", content)

        if content != original_content:
            with open(filepath, "w", encoding="utf-8") as f:
                f.write(content)
            count += 1
            print(f"Updated {filepath}")
    except Exception as e:
        pass

print(f"Updated {count} files.")

File: scripts/replace_urls.js
const fs = require('fs');
const path = require('path');

const categories = [
    'biology', 'business-startup', 'cognitive-psychology', 'conversions',
    'cooking-food', 'crypto-web3', 'education', 'employment', 'engineering',
    'environment', 'finance', 'fun-games', 'gaming', 'genetic-ancestry',
    'health-fitness', 'historical-archaeological', 'home-improvement',
    'parenting', 'personal-budgeting', 'sports-training', 'technology',
    'time-date', 'travel-adventure', 'others'
];

function walk(dir) {
    let results = [];
    const list = fs.readdirSync(dir);
    list.forEach(file => {
        file = path.join(dir, file);
        const stat = fs.statSync(file);
        if (stat && stat.isDirectory()) {
            results = results.concat(walk(file));
        } else {
            if (file.endsWith('.ts') || file.endsWith('.tsx') || file.endsWith('.md') || file.endsWith('.json')) {
                results.push(file);
            }
        }
    });
    return results;
}

const files = walk('d:/MegaCalc-Hub/src');

files.forEach(file => {
    let content = fs.readFileSync(file, 'utf8');
    let originalContent = content;

    // 1. Replace nested calculator links: `/${category.slug}/${calculator.slug}` -> `/${calculator.slug}`
    // Regex: match href="/category/calculator" or href={`/category/calculator`}
    // Watch out for subcategory e.g. /education/maths/calculator
    categories.forEach(cat => {
        // Replaces "/finance/compound-interest" -> "/compound-interest"
        const regex1 = new RegExp(`(["'\`])/${cat}/([^"'\`\\s<>]+)(["'\`])`, 'g');
        content = content.replace(regex1, (match, p1, p2, p3) => {
            // If p2 contains "maths/" we should still strip the category, wait no we strip BOTH category and subcategory!
            // The prompt says "All calculators use only `/${slug}`"
            // So "/education/maths/percentage" -> "/percentage"
            const finalSlug = p2.split('/').pop();
            return `${p1}/${finalSlug}${p3}`;
        });
    });

    // 2. Replace template literals pointing to categories to `/category/${category.slug}`
    // `href={\`/\${category.slug}\`}` -> `href={\`/category/\${category.slug}\`}`
    content = content.replace(/href=\{`\/\$\{category\.slug\}`\}/g, 'href={`/category/${category.slug}`}');
    content = content.replace(/canonical:\ `\/\$\{category\.slug\}`/g, 'canonical: `/category/${category.slug}`');

    // 3. Replace sitemap template literal URLs
    content = content.replace(/url:\ `\$\{baseUrl\}\/\$\{category\.slug\}`/g, 'url: `${baseUrl}/category/${category.slug}`');

    // 4. Schema generator template literal URLs
    content = content.replace(/"url":\ `\$\{baseUrl\}\/\$\{category\.slug\}`/g, '"url": `${baseUrl}/category/${category.slug}`');
    content = content.replace(/"item":\ `\$\{baseUrl\}\/\$\{category\.slug\}`/g, '"item": `${baseUrl}/category/${category.slug}`');
    content = content.replace(/item:\ `https:\/\/mycalculating\.com\/\$\{category\.slug\}`/g, 'item: `https://mycalculating.com/category/${category.slug}`');

    // 5. Hardcoded exact category links: href="/finance" -> href="/category/finance"
    categories.forEach(cat => {
        const regex2 = new RegExp(`href=(["'\`])/${cat}(["'\`])`, 'g');
        content = content.replace(regex2, `href=$1/category/${cat}$2`);
    });

    // 6. Fix nested dynamic strings `/${category.slug}/${calculator.slug}` in related links or breadcrumbs:
    content = content.replace(/href=\{`\/\$\{category\.slug\}\/\$\{([a-zA-Z0-9_.]+)\}`\}/g, 'href={`/${$1}`}');
    content = content.replace(/href=\{`\/\$\{category\}\/\$\{([a-zA-Z0-9_.]+)\}`\}/g, 'href={`/${$1}`}');
    // For `education/maths/...` case
    content = content.replace(/href=\{category\.slug === 'education' && calc\.subcategory === 'maths' \? \`\/education\/maths\/\$\{calc\.slug\}\` : \`\/\$\{category\.slug\}\/\$\{calc\.slug\}\`\}/g, 'href={`/${calc.slug}`}');

    // 7. Fix `app/[slug]/[...path]/page.tsx` canonical:
    content = content.replace(/canonical:\ `\/\$\{category\.slug\}\/\$\{subcategory\.slug\}`/g, 'canonical: `/category/${category.slug}/${subcategory.slug}`');

    if (content !== originalContent) {
        fs.writeFileSync(file, content, 'utf8');
        console.log(`Updated: ${file}`);
    }
});
